enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,1050 @@
+  * `message` (optional, [string][json-string]), with a human-oriented message describing the blocking reason, as described in [the blocked-edges section](#enhanced-graph-data-schema-for-blocking-edges).
+  * `matchingRules` (optional, [array][json-array]), defining conditions for deciding which clusters have the update recommended and which do not.
+    The array is ordered by decreasing precedence.
+    Consumers should walk the array in order, and stop after the first rule they can successfully evaluate.
+    If no condition can be successfully evaluated, the update should not be recommended.
+    Each entry must be an [object][json-object] with at least the following property:
+    * `type` (required, [string][json-string]), defining the type in [the condition type registry](#condition-type-registry).
+      For example, `type: promql` identifies the condition as [the `promql` type](#promql).
+    Additional properties are defined in [the cluster-condition type registry](#cluster-condition-type-registry).
+
+### Enhanced ClusterVersion representation
+
+[The ClusterVersion `status`][api-cluster-version-status] will be extended with a new `conditionalUpdates` property:
+
+```go
+// conditionalUpdates contains the list of updates that may be
+// recommended for this cluster if it meets specific required
+// conditions. Consumers interested in the set of updates that are
+// actually recommended for this cluster should use
+// availableUpdates. This list may be empty if no updates are
+// recommended, if the update service is unavailable, or if an empty
+// or invalid channel has been specified.
+// +optional
+conditionalUpdates []ConditionalUpdate `json:"conditionalUpdates,omitempty"`
+```
+
+The `availableUpdates` documentation will be adjusted to read:
+
+```go
+// availableUpdates contains updates recommended for this
+// cluster. Updates which appear in conditionalUpdates but not in
+// availableUpdates may expose this cluster to known issues. This list
+// may be empty if no updates are recommended, if the update service
+// is unavailable, or if an invalid channel has been specified.
+```
+
+The new ConditionalUpdate type will have the following schema:
+
+```go
+// ConditionalUpdate represents an update which is recommended to some
+// clusters on the version the current cluster is reconciling, but which
+// may not be recommended for the current cluster.
+// +k8s:deepcopy-gen=true
+type ConditionalUpdate struct {
+	// release is the target of the update.
+	// +required
+	release Release `json:"release"`
+
+	// risks represents the range of issues associated with
+	// updating to the target release. The cluster-version
+	// operator will evaluate all entries, and only recommend the
+	// update if there is at least one entry and all entries
+	// recommend the update.
+	// +optional
+	Risks []ConditionalUpdateRisk `json:"risks,omitempty"`
```

Comment:
I expected this to be non-optional and require at least one element.  If we have no known risks, this is not a conditional update is it?

Reply:
Right, it's not conditional.  But I don't see a lot of benefit to requiring these either, as discussed in the *Metadata to include when blocking a conditional request* section.  I do expect us to always supply at least one risk as a policy move.  But from a consumer perspective, optional vs. required here just allows you to switch between:

> Update to A is not recommended, because the upstream update service didn't give me a `risk`, so I don't know what I'm avoiding.

and:

> Update to A is not recommended, because the upstream update service didn't give me a `risk`, and that's breaking the rules of the API!

With my cluster admin hat on, both of those would be enough to understand that I needed to talk to the folks running my `upstream` update service.

Reply:
@deads2k This one is a method for us to unconditionally block edges while still providing explanations for why we're doing so which has been sorely missed in the past, to the point that we're shipping "Text Only Errata" notices whenever we block edges today which are most likely stale or inaccurate before they even reach a customer's inbox.

For all of your other comments where you feel that things are required I agree that they should be required of OCP's Cincinnati service but we're also trying to be flexible enough in the API implementation to allow other update service providers the flexibility to forego our more rigid policy. Does this make sense?

We will most likely implement linters to ensure that our curated data makes these fields mandatory.

Reply:
@sdodson why model that as a conditional update with no risks instead of a conditional update with an "AlwaysRisky" (or some similar) type?

Reply:
I like this suggestion of a risk `type: always`